File: Part 2 - Regression/Section 5 - Multiple Linear Regression/Hao_多元线性回归.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed May 29 17:41:37 2019

@author: lianghao
"""

#DataProcessing

#import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#import dataset
dataset = pd.read_csv('50_Startups.csv')
#Matrix
X = dataset.iloc[:,0:-1].values
#Factor
Y = dataset.iloc[:,4].values

#Change 'State' into dummy variables
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelEncoder_X = LabelEncoder()
#Step1: Create categories but put them in random order, which is not nessesary
X[:,3] = labelEncoder_X.fit_transform(X[:,3])
#Step2: Create categories but put them into certain orders.
from sklearn.preprocessing import OneHotEncoder
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()
#Step3: Avoid Dummy Variables Trap
X = X[:,1:]

#spliting the dataset into training set and testing set
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

#Fitting Multiple Regression into the Training Set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, Y_train)

#Predicting the Test set results
Y_pred = regressor.predict(X_test)

#Building the optimal model using Backward Elimination 反向淘汰的准备工作，加一列1，因为原本没有常数项
import statsmodels.formula.api as sm
X_train = np.append(arr = np.ones((40,1)), values = X_train, axis = 1)
#Step1: set the significant level
X_opt = X_train[:,[0,1,2,3,4,5]]
#因为重新倒入了一个标准库？
regressor_OLS = sm.OLS(endog = Y_train, exog = X_opt).fit()
regressor_OLS.summary()
#Step4: 去掉最高的
X_opt = X_train[:,[0,1,3,4,5]]
regressor_OLS = sm.OLS(endog = Y_train, exog = X_opt).fit()
regressor_OLS.summary()
#Step4: 去掉最高的
X_opt = X_train[:,[0,3,4,5]]
regressor_OLS = sm.OLS(endog = Y_train, exog = X_opt).fit()
regressor_OLS.summary()
#Step4: 去掉最高的
X_opt = X_train[:,[0,3,5]]
regressor_OLS = sm.OLS(endog = Y_train, exog = X_opt).fit()
regressor_OLS.summary()
#Step4: 去掉最高的
X_opt = X_train[:,[0,3]]
regressor_OLS = sm.OLS(endog = Y_train, exog = X_opt).fit()
regressor_OLS.summary()

File: Part 2 - Regression/Section 5 - Multiple Linear Regression/Hao_多元线性回归.R
#Data Processing

#Importing DataSet
dataset = read.csv("50_Startups.csv")

#transform categorical data
dataset$State = factor(dataset$State,
                         levels = c("New York", "California", "Florida"),
                         labels = c(1,2,3))

#splitting the dataset into training set and testing set
install.packages('caTools')
library(caTools)
#generate random number
set.seed(123)
#create split pattern based on random number generated
split = sample.split(dataset$Profit, 0.8)
#subset the source dataset into actual training set and testing set
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

#Fitting Multiple Linear Regression to the Trainning Set
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = training_set)

#Predicting the Test set Result
Y_pred = predict(regressor, newdata = testing_set)

#Building the Optimal Model using Backward Elimination
regressor = lm(formula = Profit ~.,
               data = training_set)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = training_set)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = training_set)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend,
               data = training_set)
summary(regressor)




